Issue #2166: Add 'requiredTranslations' option to Translation check

MEZk opened the pull request.

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java:
```diff
@@ -68,9 +72,17 @@
+ * <p><b>checkedLanguageCodes</b> which allows to specify language codes of
+ * translation files which must exist in project. For example, if you need
+ * to check that Japanese and French localization files are present in your project,
+ * you should set checkedLanguageCodes to "ja, fr".
```
If you'll set it like here, will the check also accept other translations? Please clarify in docs.
Also, what about ROOT translation (no code in file name)?

MEZk replied:
@mkordas 
Corrected javadoc/xdoc. Added check for default translation.
The name of default translation file begins with the base name of the ResourceBundle and ends with the .properties suffix.
FYI: https://docs.oracle.com/javase/tutorial/i18n/resbundle/propfile.html

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java:
```diff
@@ -91,6 +110,11 @@
     /**
+     * Language codes of translation files (de, pt, ja, etc).
+     */
+    private SortedSet<String> checkedLanguageCodes = Sets.newTreeSet();
```
Can we create here just empty set?

MEZk replied:
@mkordas 
Yes. why not? Sets.newTreeSet() returns a new, empty set. By default we cannot imagine what kind translations are needed.
Collections.EMPTY_SORTED_SET is not supported in Java 1.7 (marked as since 1.8).

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java:
```diff
@@ -114,10 +147,75 @@ public void finishProcessing() {
+     * @param translationFilesMap the translation files bundles organized as Map.
+     */
+    private void checkExistenceOfTranslationFiles(Map<String, Set<File>> translationFilesMap) {
+        final Set<Entry<String, Set<File>>> resourceBundles = translationFilesMap.entrySet();
```
Maybe we can inline this variable? Type for it is quite complex :)

MEZk replied:
@mkordas 
If I make it like this:

````java
private void checkExistenceOfTranslationFiles(Map<String, SortedSet<File>> translationFilesMap) {
    for (String baseName : translationFilesMap.keySet()) {
        final SortedSet<File> filesInResourceBundle = translationFilesMap.get(baseName);
    checkExistenceOfTranslationFilesInResourceBundle(filesInResourceBundle);
    }
}
````
then findbugs will rise violation.

```
[INFO] com.puppycrawl.tools.checkstyle.checks.TranslationCheck.checkExistenceOfTranslationFiles(Map) makes inefficient use of keySet iterator instead of entrySet iterator [com.puppycrawl.tools.checkstyle.checks.TranslationCheck] At TranslationCheck.java:[line 164]
```
I don't sure how to make it simpler.

mkordas replied:
Use entrySet instead keyset to iterate over map

MEZk replied:
@mkordas 
Done.

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java:
```diff
@@ -114,10 +147,75 @@ public void finishProcessing() {
+    private void checkExistenceOfTranslationFiles(Map<String, Set<File>> translationFilesMap) {
+        final Set<Entry<String, Set<File>>> resourceBundles = translationFilesMap.entrySet();
+        for (Entry<String, Set<File>> bundle : resourceBundles) {
+            final SortedSet<File> filesInResourceBundle = new TreeSet<>(bundle.getValue());
```
Can you explain why we need to create one more set here? Can't it be proper `TreeSet` from the very beginning?

MEZk replied:
@mkordas 
Is it a good idea to change
```Map<String, Set<File>> translationFilesMap``` to ```Map<String, SortedSet<File>> translationFilesMap``` and use it in the whole TranslationCheck class?

mkordas replied:
Yes, I think it's good idea

MEZk replied:
@mkordas 
Done.

mkordas commented on src/main/resources/com/puppycrawl/tools/checkstyle/messages_ja.properties:
```diff
@@ -0,0 +1,15 @@
+PackageObjectFactory.unableToInstantiateExceptionMessage=''{0}'' クラスをインスタンス化することができません、\
+  それは としてインスタンス化することもできません {1}. \
+  そのクラス名が正規名として指定されている再確認または短い名前の使用 \
+  http://checkstyle.sourceforge.net/config.html#Packages を設定する方法をお読みください。\
```
We need to escape `#` here, because otherwise it starts comment. Please also recheck other property files

MEZk replied:
@mkordas 
Done.

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java:
```diff
@@ -68,9 +72,17 @@
+ * translation files which must exist in project. For example, if you need
+ * to check that Japanese and French localization files are present in your project,
+ * you should set checkedLanguageCodes to "ja, fr".
+ * Default value is <b>{}</b> which means that the existence of
```
What do you mean by `{}` here? It is neither empty string nor empty set.

MEZk replied:
@mkordas 
I mean empty string set. Corrected javadoc/xdoc.

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:
```diff
@@ -567,4 +567,46 @@ public void testListFilesDirectoryWithNull() throws Exception {
     }
+
+    @Test
+    public void testTranslationCheckMissinTranslationFiles() throws Exception {
```
Why do we have this test in `MainTest`? It should belong to `TranslationCheckTest`

mkordas replied:
`Missin` -> `Missing`

MEZk replied:
@mkordas 
Movet to TranslationCheckTest.
Corrected typo.

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:
```diff
@@ -567,4 +567,46 @@ public void testListFilesDirectoryWithNull() throws Exception {
+    @Test
+    public void testTranslationCheckMissinTranslationFiles() throws Exception {
+        exit.expectSystemExitWithStatus(6);
+        final String platforIndependentPath = "src" + File.separator + "test"
```
You can use `String.format` here to not repeat `File.separator` so many times

mkordas commented:
@MEZk 
I'm done with my code review, just two more things:

1. `checkedLanguageCodes` will become our API, so we should choose this name carefully. Are you sure about this name? Maybe you can propose some other options and we'll do some voting? My proposition would be `requiredTranslations`.
2. We should correct grammar in commit message. I propose `Issue #2166: Add 'checkedLanguageCodes' option to Translation check` instead of `Issue #2166: Add check of existence of translation files for TranslationCheck`.

MEZk commented:
@mkordas 
Thank you for so detailed review, I'll provide changes ASAP.

MEZk commented:
@mkordas 
>My proposition would be requiredTranslations.

I support your idea.

2) Corrected.

mkordas commented:
@MEZk 
I've provided detailed review inline in the commits. We're much closer to the finish now. Please should if you disagree with something.

MEZk commented:
@mkordas 
>Link to ISO 639-1 language codes should be part of some sentence

Corrected.

>This Javadoc is the same as for field above - they should be different as they describe different fields I believe.

Corrected.

>We don't need to allocate any memory for TreeSet here, so I propose to use ImmutableSortedSet.of(); instead.

Replaced with ImmutableSortedSet.of()

>This is duplicated code: ... Please extract it to a method

Done.

>Duplicated operation: .substring(baseName.lastIndexOf(File.separator) + 1).
Please extract it to a private static method.

Done.

>I think is missing should be used instead of is missed.

Corrected in all translation files.

>Whenever we will change anything in Main class this test will start failing.
Can't we create just unit test for TranslationCheck? Please explain why it is not possible to avoid using Main for this test.

My bad. We can create just unit test for TranslationCheck.

>Do we really need to configure charset, 'severityandfileExtensions` for this test? It's not unit test anymore...

I've deleted the file, since it is not necessary anymore.

>1) It's link to page named ISO639-2, shouldn't it point to some page with just ISO639-1 codes to avoid confusion?
2) It should be part of some sentence
3) Please specify impact of this propery to default translation 

1) Corrected.
2) Yes, corrected.
3) Done.

mkordas commented on src/xdocs/config_misc.xml:
```diff
@@ -328,6 +328,20 @@ messages.properties: Key 'ok' missing.
+             <td>
+                 Allows to specify language codes of required translations which must exist in project.
+                 Language code is composed of the lowercase, two-letter codes as defined by
+                 <a href="http://www.fatbellyman.com/webstuff/language_codes_639-1/">ISO 639-1</a>
```
very minor: missing dot at the end of sentence

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java:
```diff
@@ -120,10 +120,53 @@ public void testLogIOException() throws Exception {
+    }
+
+    @Test
+    public void testTranslationFileAreMissing() throws Exception {
```
minor: testTranslationFile**s**AreMissing

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java:
```diff
@@ -120,10 +120,53 @@ public void testLogIOException() throws Exception {
+    @Test
+    public void testDefaultTranslationFileIsMissing() throws Exception {
+        final DefaultConfiguration checkConfig = createCheckConfig(TranslationCheck.class);
+        checkConfig.addAttribute("requiredTranslations", "ja, de, ja");
```
this test could include more complicated string set like ",ja,de,, ja," just to have full functionality coverage

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java:
```diff
@@ -120,10 +120,53 @@ public void testLogIOException() throws Exception {
         Method logIOException = check.getClass().getDeclaredMethod("logIOException",
-                IOException.class,
-                File.class);
+            IOException.class,
```
looks like unnecessary changes in scope of this PR

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java:
```diff
@@ -112,12 +147,130 @@ protected void processFiltered(File file, List<String> lines) {
+                }
+            }
+
+            if (translationFileIsMissed) {
```
`isMissing` is more common I think

mkordas commented:
@MEZk, thanks a lot! I've added a few final, very minor comments to PR. 
@romani, I think you can start your code review and merge if no addtional issues will be found

MEZk commented:
@mkordas 
>very minor: missing dot at the end of sentence

Corrected.

>minor: testTranslationFilesAreMissing

Corrected.

>this test could include more complicated string set like ,ja,de,, ja, just to have full functionality coverage

Added more complicated string set like ,ja,de,, ja.

>looks like unnecessary changes in scope of this PR

Corrected.

>isMissing is more common I think

Renamed.

P.S. Thanks. I appreciate your reviews.

romani commented:
in second commit I still see "#" , please search for - "http://checkstyle.sourceforge.net/config.html\#Packages" , other comments will be later.

MEZk commented:
@romani 

```http://checkstyle.sourceforge.net/config.html#Packages``` will be produced to user. ]
See [Test](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java#L189)

romani commented:
ok, I miss escape symbol.

MEZk commented:
@mkordas , @romani 

Added new option **requiredTranslations***
Added new option **translationsNotRequiredForBundleNamePattern**

romani commented:
> translationsNotRequiredForBundleNamePattern

No, please go from positive logic and search only for "messages_*.properties" as format and do not touch any other property files.

MEZk commented:
@romani 
Please explain why we should check only   "messages_*.properties". Oracle doesn't prohibit use of other names for translations bundles. See https://docs.oracle.com/javase/tutorial/i18n/resbundle/propfile.html

Or it is common practise to call translations bundles as "messages_*.properties" ?

romani commented:
Let me investigate this, but main point is that *.properties files is general format to keep some string values , there is no limitations to use it for messages only. Right now more and more project try to use xml, json, yml for this still there are should be number of example where properties are still in use.

romani commented:
@MEZk 

can you move translationsNotRequiredForBundleNamePattern to separate PR  ? to let other changes be applied.

MEZk commented:
@romani 
Without translationsNotRequiredForBundleNamePattern we should provide translations for checkstyle-ant-task_es.properties

MEZk commented:
@romani 
For now, I removed code connected with ```translationsNotRequiredForBundleNamePattern```
I'll create PR after you merge changes to avoid merge conflicts.

romani commented:
Create an issue for this and keep it in you branch on remote

mkordas commented:
@MEZk, how does it work right now? I see green builds without updating Checkstyle Ant Task properties.

MEZk commented:
@mkordas 
See https://github.com/MEZk/checkstyle/commit/8a6af5a2cfa101e186bb158ab4e013858eab0ee0#diff-bfaac004d98e5f0f13a9b478ac5797cfR77

mkordas commented:
Thanks, now I see. I think this PR is ready to merge. What dou you think @romani?

romani commented:
some of items are not related changes , but lets consider them too, as we do serious update to Check

1)
> final Map<String, SortedSet<File>> propFilesMap =

Why not a 
http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Multimap.html
?
Why do we need SortedTree ? only to get default as first element ? I might miss smth , but sorted collection is too much for this logic.

2)
> if (!requiredTranslations.isEmpty()) {

no need to do this. remove that check and code become clear, and you will not do performance changes as "checkExistenceOfTranslationFiles" will not iterate and skip processing.

3)
> checkExistenceOfTranslationFilesInResourceBundle

name is too big. Please remove "OfTranslationFile" , as whole Check is about translations files

4)
> isMissing = !currentFileName.contains('_' + languageCode + ".");
> return String.format("%s_%s.properties",

"_" is hardcoded while we have option for it ?
please create UTs that would fail on this code and basenameSeparator is equal "-",  please attach links to that UTs when you put a "Done" on this point.
please  more ".properties" to variable.

5)
> private static String extractPropertyIdentifier(final File file,
            final String basenameSeparator) {

please remove final modifiers, do not pollute a code. I know that this decision is inconsistent with demanding "final" on local variables .... one day I will rethink its necessity to code, not now.

MEZk commented:
@romani 
4) It is the ending format of property file name: [base name] + "_" + [language code] + [dot] + [extension].
For example:
```
messages-test + _ + ja + . + properties
```

or
````
messages + _ + ja + . + properties
````

See para 2: https://docs.oracle.com/javase/tutorial/i18n/resbundle/propfile.html
>the base name is followed by the language code

But yes, ```.properties``` should be replaced with **extension** option from base class.

MEZk commented:
@romani 
1) 
>Why do we need SortedTree ? only to get default as first element ? I might miss smth , but sorted collection is too much for this logic.

Otherwise we should iterate though each element of the Set and check whether it is a default translation. Maybe we can use SortedSet just inside ```checkExistenceOfTranslationFilesInResourceBundle``` or try to change logic?

romani commented:
> Maybe we can use SortedSet just inside checkExistenceOfTranslationFilesInResourceBundle or try to change logic?

It is make sense to use sorted collection only when you repetitive search over it few times. But our logic do search only ones for default - that is simple search that is required only ones. 

> So, It seams to me that basenameSeparator should be removed from check's options!

Just take a look at history (blame) and you will see a reason why it was added.

6)
> 
```
            boolean isMissing = false;
            for (File currentFile : resourceBundle) {
                final String currentFileName = currentFile.getName();
                isMissing = !currentFileName.contains('_' + languageCode + ".");
                if (!isMissing) {
                    break;
                }
            }
```

make is as function and name it let code read clearly.

MEZk commented:
@romani 
Now I see why it is needed.
Are you ok with the following:

1) If we use basenameSeparator, a base name of resource bundle is right before separator.
For example:
For ```app-stage.properties``` base name is ```app```.
For ```app_stage.properties``` base name is ```app```.
For ```app-stage_de.properties``` base name is ```app```.
Because right now the base name is treated as it it described in the above examples.

2) The check looks only for messages bundles which names contain the word 'messages'.

romani commented:
> 1) If we use basenameSeparator, a base name of resource bundle is right before separator.

all cases are right.

> 2) The check looks only for messages bundles which names contain the word 'messages' when it validates an existence of translations.

not sure why we need this decision in this PR, you already moved that to separate issue. Please explain.

MEZk commented:
@romani 
1) Map has been replaced with SetMultimap from Guava. SortedSet has been replaced with Set.

2)
 >no need to do this. remove that check and code become clear, and you will not do performance changes as "checkExistenceOfTranslationFiles" will not iterate and skip processing.

Done. 

3) I think ```checkExistenceOfDefaultTranslation(filesInBundle)``` and  ```checkExistenceOfRequiredTranslations(filesInBundle)``` are better names for the checks for now. Do you agree? See [checkExistenceOfTranslations](https://github.com/MEZk/checkstyle/commit/fa332a8decb58681505e46525b6a6d5344cf2786#diff-bfaac004d98e5f0f13a9b478ac5797cfR163)

4) UTs: [testBaseNameSeparatorDefaultTranslationIsMissing](https://github.com/MEZk/checkstyle/commit/fa332a8decb58681505e46525b6a6d5344cf2786#diff-42885903150bfedec158bb760f2fe753R174) and [testBaseNameSeparatorTranslationsAreMissing](https://github.com/MEZk/checkstyle/commit/fa332a8decb58681505e46525b6a6d5344cf2786#diff-42885903150bfedec158bb760f2fe753R195)

5) Done

>not sure why we need this decision in this PR, you already moved that to separate issue. Please explain.

Just to make sure that the check works fine without new option.

**In addition:**

null check was removed from  https://github.com/MEZk/checkstyle/commit/fa332a8decb58681505e46525b6a6d5344cf2786#diff-bfaac004d98e5f0f13a9b478ac5797cfL174 since SetMultimap.get(key) always returns a non-null, possibly empty collection. See [guava doc](https://github.com/google/guava/wiki/NewCollectionTypesExplained#multimap)

romani commented:
please recheck this on JDK that Travis use - https://travis-ci.org/checkstyle/checkstyle/jobs/81554715#L103, latest jdk8 is http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html

"java version "1.8.0_31" vs "8u60"

mvn version is "3.2.5" https://travis-ci.org/checkstyle/checkstyle/jobs/81554715#L59

MEZk commented:
@romani 
With maven 3.2.5.

|     | java 1.8.0.60| jav 1.8.0.31|
|---:|-----:|-----:|
|**mvn clean verify**| build success| build success|
|**mvn clean cobertura:clean cobertura:check**|build success | build success|
Strange.
 And AppVeyor uses java 1.8.0.45 and maven 3.2.5.

romani commented:
at this issue nuance was in missed "clean"

MEZk commented:
@romani 
The same problem

romani commented:
Works fine on my laptop, and other linux boxes
```
$ mvn -version
Apache Maven 3.2.2 (45f7c06d68e745d05611f7fd14efb6594181933e; 2014-06-17T06:51:42-07:00)
Maven home: /opt/maven/default
Java version: 1.8.0_45, vendor: Oracle Corporation
Java home: /opt/jdk1.8.0_45/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.13.0-63-generic", arch: "amd64", family: "unix"
```

I presume that you run out of ideas and need help , right ?

MEZk commented:
@romani 
Yes, I cannot imagine why Travis CI fails the build.

romani commented:
I need your update to be part of 6.11:
- please create separate issue to resolve this coverage problem
- add additional commit to this PR and add exclude for coverage for this Check.

romani commented:
Why you exclude it completely ? Put a certain covrrage on which travis is agree.

MEZk commented:
@romani 
>please create separate issue to resolve this coverage problem

Done. See https://github.com/checkstyle/checkstyle/issues/2243 .

>add additional commit to this PR and add exclude for coverage for this Check.
Why you exclude it completely ? Put a certain covrrage on which travis is agree.

Done.

romani commented:
merged as FF
